neuralmonkey/decoders/sequence_labeler.py

```diff
@@ -83,6 +83,11 @@ def train_targets(self) -> tf.Tensor:
     def train_mask(self) -> tf.Tensor:
         return sentence_mask(self.train_targets)
 
+    @tensor
+    def output_mask(self) -> tf.Tensor:
+        return tf.where(
+            self.train_mode, self.train_mask, self.input_mask)
```

Comment:
Když bude runtime režim, tak tohle neprojde, když nebudeš mít referenci.

Reply:
No tak to necheme, pouštět labelátor bez reference potřebujeme. Třeba v neautoregresivním překladu. V tom případě bych teda vracel input mask a při počítání costu controloval jestli je train mask stejná jako input mask.

Reply:
For je v tom, ze input mask se bere z enkoderu a ten enkoduje vsechny tokeny, train_mask ti ma zakryt ty tokeny, ktere jsi ve vstupu nemenil (a tedy te nezajima, co predikujes u ostatnich tokenu)

Reply:
vymenil jsem tf.where za tf.cond(..., lambda: self.train_mask, ...). Pri neuralmonkey-run bude self.train_mode vzdy False, takze se nikdy nezavola true_fn vetev; tensory z ni by se ani nemely vytvaret a tedy by tf nemel nadavat na prazdny placeholder
https://www.tensorflow.org/api_docs/python/tf/cond

Zatim jsem to netestoval, pokud to nekdo testne, dejte mi pls feedback

Reply:
Pro tensorflow 1.12:
```python
In [1]: import tensorflow as tf

In [2]: a = tf.placeholder(dtype=tf.int32)

In [3]: b = tf.constant(4)

In [4]: c = tf.cond(tf.constant(False), lambda: a, lambda: b)

In [5]: s = tf.Session()

In [7]: s.run(c)
InvalidArgumentError: You must feed a value for placeholder tensor 'Placeholder' with dtype int32

In [8]: s.run(c, feed_dict={a: 3})
Out[8]: 4
```

Reply:
Ale to je proto, ze ten tensor vytvoris mimo tu lambda funkci.
Zkus si:
`
c = tf.cond(tf.constant(False), lambda: tf.placeholder(dtype=tf.int32), lambda: tf.constant(4))
tf.Session().run(c)
`

Reply:
aha, zajímavý.. nicméně tady je self.train_mask i self.input_mask už vytvořený, ne? respektive, možná neni, ale je to docela křehkej předpoklad.

Reply:
pri runtime, se self.train_mask nema kde jinde volat, takze vytvoreny nebude. Potencialne se muze vytvorit jen v te lambda funkci, ale tam by se to nemelo dostat.

Stale rikam, ze je potreba to vyzkouset, ale sam jsem se k tomu jeste nedostal.